### ISSUE TITLE
[code_search] Fix the loss function

### ISSUE THREAD

**Issue by jlewi** [kubeflow/examples#259]
There was discussion in slack about trying different loss functions. I pasted the conversation below but here are the ideas mentioned

1. @activatedgeek suggested using fully connected layer with one sigmoid output
1. @hamelsmu suggested using cosine distance, followed by tanh, use a logits function that doesn't involve logits e.g sparse_categorical_crossentropy (details below).

Sanyam Kapoor [6 hours ago]
I think instead of this, a better approach would have been to just have a fully-connected layer with one sigmoid output. The normalized dot-product is kind of limiting the range of possibilities. I put this in my mental backlog because any change to the model would stay isolated architecturally. (edited)

Hamel [3 hours ago]
ok what about this.  re: Cosine Distance, Lets compute cosine distance as just the normalized dot product (leave out the bit where you subtract this from 1), you will end up with a number between [-1, 1]  -- (right now you are getting a number between [0, 2] (edited)

Hamel [3 hours ago]
After that, we want to transform this number that is in a range between [-1, 1] to something that is bounded by [0,1] --> to do this pass the cosine distance you computed in the previous step  through a `tanh` activation function (edited)

Hamel [3 hours ago]
Lastly, because `tf.nn.sigmoid_cross_entropy_with_logits` expects logits in the `logits` argument, use a different loss function here that doesn't expect logits (logits are un-normalized values that have not been passed through a softmax) (edited)

Hamel [3 hours ago]
you can try to use this loss function in TF that doesn't expect logits and will instead take your normalized predictions that are between [0,1]: `from tf.keras.objectives import sparse_categorical_crossentropy` (edited)

Hamel [3 hours ago]
this way, you can feed the raw values from the tanh that are values that are bounded between [0,1] and compare those to target variable which is binary 0/1

Hamel [3 hours ago]
Maybe the logits part may not matter, but I think weird things happen when you pass normalized values to `sigmoid_cross_entropy_with_logits` instead of raw values (I forget the math right now something about sigmoid transformation being sensitive to values < 1), also this might help readability of the thing. (edited)

Hamel [3 hours ago]
Let me know if these comments are hard to follow I can also make annotations in GitHub if helpful

Christopher [3 hours ago]
Updated notebook here that should be useful for playing around with these things https://github.com/cwbeitel/tk/blob/master/docs/code_search_debug.ipynb, including playing around with a cosine similarity loss that was experimented with as part of slicenet (?) https://github.com/tensorflow/tensor2tensor/blob/660d49fc7a20025c1a8ae4448ad92022f042f02b/tensor2tensor/models/slicenet.py#L143
GitHub
cwbeitel/tk
Contribute to cwbeitel/tk development by creating an account on GitHub.
 
 GitHub
tensorflow/tensor2tensor
Library of deep learning models and datasets designed to make deep learning more accessible and accelerate ML research. - tensorflow/tensor2tensor
 

Christopher [3 hours ago]
Training and prediction runs without errors  (for a batch size of 1, which I think is a tf eager thing), requires tf v1.6 because of a bug eager in 1.7.

Christopher [3 hours ago]
That slicenet loss is basically summing the distance to all non-self elements in the batch iiuc

Christopher [3 hours ago]
That loss is ignoring the distance to self which we can accomplish by constructing the right matrix of labels and doing cross-entropy or separate out the diagonal from non-diagonal and compute partial sums separately

Christopher [2 hours ago]
Also correcting what I said above putting it on [0,2] instead of [0,1] before the log will mean the input for the sigmoid will be centered on 0 (log(1) = 0) which is what you want.

Christopher [2 hours ago]
@Hamel I think I have a general idea of what you’re saying. I think we’ll answer some of these questions by just trying a bunch of experiments and seeing what happens.

**Comment by hamelsmu** [comment]
The way I see it, the current implementation of using the `cosine distance`, which is bounded between [0,2] as the logit **represents a bug.**  This is because logits are un-normalized values before being passed to softmax.  

Here are some of the reasons why I believe this is a bug:

1.  typically as input into tf.nn.sparse_softmax_cross_entropy_with_logits for the `logits` parameter, you would provide a 2-D tensor representing (batch size, # of classes).  The value being passed here appears to be a 2-D tensor that represents (batch size, 1), where we are passing single values in the second dimension.   the loss function `tf.nn.sparse_softmax_cross_entropy_with_logits` expects the  `logits` parameter to be the raw predictions of your model.  For example, let's say you have batch size = 3 and 2 possible classes, so a tensor of shape of (3, 2):

Logits are this:

[[8, 2]
[7, 3]
[6, 4]]

Behind the scenes, Tensor flow passes this through a softmax layer to squash these between [0,1] so will become this:

[[ 0.8, 0.2],
 [0.7, 0.3],
 [0.6, 0.4]]

2.  Even if I am misunderstanding above, it doesn't make sense to me to treat cosine distance [0, 2] as a logit.  this kind of means that these values [0,2] will be transformed to be on a scale of [0, 1] where cosine distance of 0 --> 0 and cosine distance 2 --> 1.   However, in this case, a value of 2 means that the vectors are very far apart so this would lead the exact opposite of the desired behavior.  However, because of #1 I don't think the point I'm making here is even important.  

As an aside, look for a binary cross entropy function instead of sparse cross entropy.  The labels are not sparse in this example as they are only binary.  We might be able to save a little bit of work (and maybe some memory) by just having to keep track of the predicted probability of the positive class.  

Please let me know if I am misunderstanding something.  This is my analysis of reading the code and the discussion but I could be missing something.  I hope this helpful.  Thanks, I find it very helpful and have learned a lot by lurking in these repos.

**Comment by hamelsmu** [comment]
Actually I made a mistake in this comment editing it now.

**Comment by hamelsmu** [comment]
Ok I updated my explanation to correct a detail that I overlooked, however, my recommendation is very similar.

**Comment by cwbeitel** [comment]
Yeah I agree, we should be using cosine similarity not cosine distance. I misunderstood initially (equated the two) and thought that's what we were using which is maybe why it was confusing what I was saying to build on @jlewi's idea of using a log - we can apply the log to *cosine similarity* values that have been shifted by one to be on [0, 2] instead of [-1, 1]. The result of this will be centered on 0 since log(1) is zero and application of the sigmoid will yield values on [0,1] (congruent with @hamelsmu's point that 1's coming out of the sigmoid should correspond to a "match" since that's how we're encoding down-stream).

We can also try just scaling the cosine similarity from [-1,1] to [-k,k] before applying the sigmoid.

Also tanh has a range of [-1,1] so if your input to that has a range of [-1,1] you're not converting it to [0,1] - for the tanh to have a range of [0,1] its input would need to be on [0,N].

I think it's also worth considering the rank loss I linked above, here's that [again](https://github.com/tensorflow/tensor2tensor/blob/660d49fc7a20025c1a8ae4448ad92022f042f02b/tensor2tensor/models/slicenet.py#L143), which I basically understand to be summing 1 - the cosine similarity off the diagonal together with the cosine similarity along the diagonal, saying, as they comment in the docstring, that you should be more similar to your own targets than to others. If I understand it correctly it could be simplified as follows:

```python

# compute cosine distance scores
docstr_emb = tf.nn.l2_normalize(docstr_emb, 1)
code_emb = tf.nn.l2_normalize(code_emb, 1)
scores = tf.matmul(docstr_emb, tf.transpose(code_emb)) 

# get scores that refer to two embeddings that should correspond
diagonal = tf.diag_part(scores)

# Sum the values off the diagonal with 1 - values on diagonal, k >=0, maybe 2
# Will have values on range [-B^2, B^2] for batch size B.
score = tf.reduce_mean(scores - k*diagonal)

# k can be a function of b to control weighting of importance of self vs. 
# non-self distances.

```

In the above, values off the diagonal should be low or negative and values on the diagonal should be high, so taking 1- that on the diagonal and summing the two should have a low total value for good embedding functions.

**Comment by cwbeitel** [comment]
Ugh again! Made one edit, changed "compute cosine distance scores" to "compute cosine similarity scores" in the first code comment.

**Comment by hamelsmu** [comment]
@cwbeitel sorry, I meant `sigmoid` activation, not `tanh`

**Comment by hamelsmu** [comment]
And yeah its not perfect since the values going in are bounded between [-1, 1] so the values coming out will be bounded at ~  [.25, .75] instead of [0 , 1] but should still work, I think the rank loss you suggest could be good too

**Comment by cwbeitel** [comment]
Hmm yeah I agree about the [0.25,0.75] thing and that's why I was proposing to try scaling out to [-k,k] to more broadly span the domain of the sigmoid but I take that back - we want to make sure our distance measure is incentivizing the learning of embeddings relative to the nearest-neighbor lookup we'll be doing at search time. E.g. if we're doing lookup within a radius of 0.1 then we don't want our sigmoid to be squashing our ability to distinguish distances at that boundary. If we scaled out to [-k,k] it would be as if we're building a classifier for whether things are pointed in the same or different general direction.

So we could skip the sigmoid or shift and squash it's activation point to be right at the decision boundary. Or shift the input, i.e. if the decision boundary is at a distance of 0.1, we could just subtract 0.9 from the cosine similarity values before applying the sigmoid.

**Comment by cwbeitel** [comment]
In the case where we do this kind of shift I would then advocate also scaling out from [-1,1] to [-k,k] which would put the shift at -k*(1-d) for a lookup distance of d.

**Comment by cwbeitel** [comment]
That was confusing about shifting and "squashing" the sigmoid's activation point - that's what we're accomplishing by shifting and stretching the input, i.e. re-positioning the decision boundary and making the distinction more precise.

**Comment by hamelsmu** [comment]
Ok I must admit that it feels like we are in the headspace of kind of making this up.  I feel like someone must have tackled this before.  Will try to look at some papers to see what other people do in this paradigm (maybe you already have — what are others doing?).

**Comment by hamelsmu** [comment]
However, if we are comfortable making up our own scheme (why not) here is a way to do it.  

Right now the output of the sigmoid activation discussed above gives you a number on the range [0.25, 0.75]  We can stretch this value to [0, 1] by using the following transformation:  `f(x)  = 2x - 0.5`.   

(.25 * 2) - .5 = 0
(.75 * 2) - .5 = 1

I am not sure this transformation makes any difference in the model's ability to learn the proper weights, as we are just rescaling the input values a bit but it should still work just fine without doing this transformation.

**Comment by cwbeitel** [comment]
That's also a good idea and I agree it's worth trying.

The sigmoid is a continuous function so you can just shift and squash it by altering the domain continuously. If you multiply by k your similarities will still be centered on 0 but for sufficiently large k sig(k) ~ 1 and sig(-k) ~ -1. E.g. for k=3 a similarity of 1 would be sig(kx)=0.95.

With the shift we'd be computing f(x) = sig(k(x-(1-d))) which for k=20 take values:

x=1, f=0.881
x=0.9, f=0.500
x=0.8, f=0.119
x=0.7, f=0.017
x=0, f ~ 0
x=-1, f ~ 0

That's an extreme example, you may not want to squash it that much and but that illustrates the point about how you can emphasize the decision boundary. For a smaller k you can still scale the output of the sigmoid as you were mentioning if you want to.

We could also consider other losses that impose the requirement that arithmetic performed in the space has a sort of self-consistency ... which maybe moves more in the direction of a curiosity but is neat to consider! - e.g. we could apply something like [markov clustering](https://micans.org/mcl/) (there might be a differentiable analog of that algorithm) to the doc string and code embedding all-by-all similarity matrices independently, then compute the dissimilarity of the resulting matrices, in effect saying that we'd like the overall structure of the doc string space to be similar to the overall structure of the code space.

**Comment by hamelsmu** [comment]
Also, arranging it in the way I described above still incentivizes the network to force the cosine proximity of the embeddings towards -1 when the (code, query) is not similar and towards 1 when the (code, query) is about the same thing.   So I think we don't have to worry if the predictions take up the entire scale of [0, 1]  ..... this is just a hack anyways to learn embeddings such that similar ones are pushed together in cosine space.   Thoughts?

**Comment by cwbeitel** [comment]
Oh you posted your last two before I refreshed. I have not done a lit review, that's a good idea, only was inspired toward that last thing by papers showing you can do arithmetic in embedding spaces (which is something you might want to actually incentivize in your loss).

It might indeed work fine without doing the transformations or getting complicated, I agree.

**Comment by cwbeitel** [comment]
I agree that it incentivizes the optimization in the right direction but I think the vanilla sigmoid isn't emphasizing that we aren't simply trying to tell the difference between close and far, we're trying to distinguish between a specific small region around a query and everything not in that region.

Someone at github should suggest a feature of live-refreshing comments on issues 😉

**Comment by hamelsmu** [comment]
I know man, sometimes github issues drive me nuts.  

Question about this:

>the vanilla sigmoid isn't emphasizing that we aren't simply trying to tell the difference between close and far, we're trying to distinguish between a specific small region around a query and everything not in that region. 

Since we are using cross entropy loss, the loss function penalizes us more the more "incorrect" our predicted probability is.  For example if we predict .65 and the true label is 1 then we will be penalized more than if we predict .75.   Because of the continuous nature of cross entropy loss, I would think that this network can learn the difference b/w close and far and is not just binary in nature.   

However, maybe I'm missing something?  Whats your perspective?

**Comment by cwbeitel** [comment]
With that k=20 example above that's like saying "things that correspond should have very high distances, above 0.9 or so" and "measures of distance for things that are opposites or not close don't matter". Idk if we really want to say the last but that's the intuition.

**Comment by cwbeitel** [comment]
I agree it might work just fine without the fancy loss. But ultimately the loss expresses what you want to accomplish and the easier you can make the optimization the better. Given infinite optimization time the difference wouldn't matter, I think we're considering this in terms of simplifying the challenge of the opt.

**Comment by hamelsmu** [comment]
Ok.  By the way I believe Lukas at Google recommended hinge loss when advising us on this project.

**Comment by activatedgeek** [comment]
There's one more to the loss part here. We also need to mask the padded string so that the gradients don't flow in from there.

**Comment by cwbeitel** [comment]
Yeah I think what they were doing in the [rank_loss code](https://github.com/tensorflow/tensor2tensor/blob/660d49fc7a20025c1a8ae4448ad92022f042f02b/tensor2tensor/models/slicenet.py#L152) is like the code I provided above plus hinge?

@activatedgeek Can you elaborate, I don't understand.

**Comment by cwbeitel** [comment]
You could also compare the similarity matrix with its transpose, i.e. comparing d(Di, Cj) with d(Cj, Ci which in a consistent embedding space would be the same.

**Comment by hamelsmu** [comment]
Rank loss probably works fine too, I haven’t looked into that to be honest.  Thanks for pointing to the T2T code, I’ll take a look

**Comment by activatedgeek** [comment]
Thank you @hamelsmu for pointing out the bad cosine distance. It was a clear oversight on my part.

And also the point about learning the fact that similar things need to close AND dissimilar things need to be far is extremely important here.

I am currently trying two things:

1. Rank Loss as @cwbeitel suggested
2. Contrastive Loss (http://yann.lecun.com/exdb/publis/pdf/hadsell-chopra-lecun-06.pdf)

Quick look at both formulas, it appears that both use the idea of margin.

**Comment by activatedgeek** [comment]
@hamelsmu I modified the code to do a simple sigmoid after unnormalized dot product (See [here](https://github.com/activatedgeek/examples/blob/update-similarity-transformer/code_search/src/code_search/t2t/similarity_transformer.py)).

The loss stays constant after some time at ~0.5 even with this update. It appears that the classifier is completely confused (always giving probabilities close to 0.5).

**Comment by cwbeitel** [comment]
Various losses [here](https://github.com/cwbeitel/tk/blob/master/tk/models/similarity_transformer.py#L45) converge but so far I haven't evaluated whether that's meaningful qualitatively.

Those losses have nicknames in the following notebooks: "kfnet" (kubeflow_similarity_cost), "simple_cd" (simple_similarity_cost), and "slicenet" (slicenet_similarity_cost).

Notebooks [here](https://github.com/cwbeitel/tk/blob/master/docs/cs_fix_loss.ipynb) and [here](https://github.com/cwbeitel/tk/blob/master/docs/cs_fix_loss2.ipynb).

The hinge cosine similarity loss poached from slicenet seems to make progress, here's a plot showing it relative to a previous non-functioning verison of the kfnet and simple_cd losses.

<img width="690" alt="slicenet" src="https://user-images.githubusercontent.com/2348401/46926648-70ff3000-cfe6-11e8-836a-2c372a730684.png">

For the kfnet applying the scale factor I was suggesting above to shift the input for the sigmoid from [-1,1] to [-k,k] seems to be helpful, here with k=20 in orange:

<img width="693" alt="kfnet" src="https://user-images.githubusercontent.com/2348401/46926638-6b094f00-cfe6-11e8-8177-1a24e0bb2e5c.png">

For the simple_cd loss you can see the original stagnant version in the first figure which included a last line like so: `tf.reduce_mean(cosine_similarity - k*self_cosine_similarity)`. Turning up k yielded 

Work on the simple_cd loss is a little unclear b/c I wasn't tracking hparams well but the initial version was stagnant (shown in the first figure), and involved a last line like so: `tf.reduce_mean(cosine_similarity - k*self_cosine_similarity)`. Presumably this was an issue because non-self similarity values would swamp self values. So initially to compensate for this I turned up k to 250-1000 yielded a loss curve very similar to the orange line in the second plot above with the idea being to compensate for the B^2 - B greater number of non-self associations (for batch size B) - but it make a lot more sense to accomplish this by just changing that to `tf.reduce_mean(cosine_similarity) - k*tf.reduce_mean(self_cosine_similarity)` which yielded the following:

<img width="692" alt="simplecdsepsum" src="https://user-images.githubusercontent.com/2348401/46926874-81fc7100-cfe7-11e8-8bc4-862225c3a065.png">

I'm guessing that is just an issue of playing with the k value.

So the take-away here is there are various losses that seem to converge but unknown yet whether that's meaningful.

**Comment by cwbeitel** [comment]
Additional experiments and qualitative analysis notebooks [here](https://github.com/cwbeitel/tk/blob/7d7a1a777a2cb1a47a86a2b53963d9896a7ea437/docs/cs_fix_loss3.ipynb) and [here](https://github.com/cwbeitel/tk/blob/7d7a1a777a2cb1a47a86a2b53963d9896a7ea437/docs/cs_local_search.ipynb).

Something you'll note in the notebooks is that I started experimenting with a transpose similarity loss. Intuitively we not only expect to have high self-similarity within doc-string and code pairs but we also expect each to relate to the other in the same way (i.e. given a matrix of doc string / code embedding similarities, we expect row i to be similar to column i, not just to have the value of (i,i) be relatively high.)

<img width="542" alt="heat" src="https://user-images.githubusercontent.com/2348401/47738134-b9962a80-dc2f-11e8-9a1b-3ff38f67e7a8.png">

There are two such measures I've been experimenting with (1) tf.reduce_mean(tf.abs(cs - tf.transpose(cs))) and (2) 0.5*(1 - tf.reduce_mean(_cosine_similarity(cs, tf.transpose(cs)))). The former is probably overly simple and it would probably be better to try using something like the euclidean distance. Note in the latter we're using the cosine distance (again, i.e. between cosine distance vectors).

Notably some losses (namely the "simple_cd" or balanced cosine similarity loss) actually increase the TCS as well as show much greater overall variance in these. The longest run shown below (red) is with the hinge loss from the t2t code base we've been discussing. As you can see TCS stays low and TS and the error-correction training loss continue to make progress over the course of the 200k steps of training.

<img width="1028" alt="ts" src="https://user-images.githubusercontent.com/2348401/47738135-b9962a80-dc2f-11e8-8b43-ab3d57f2e7bd.png">

<img width="1033" alt="tcs" src="https://user-images.githubusercontent.com/2348401/47738136-b9962a80-dc2f-11e8-8807-c1b4abbbc3cc.png">

<img width="1031" alt="training" src="https://user-images.githubusercontent.com/2348401/47738137-ba2ec100-dc2f-11e8-9ddd-440d1fb36371.png">

But when examining this in terms of the quality of searches these appear to still be quite low, see [notebook](https://github.com/cwbeitel/tk/blob/7d7a1a777a2cb1a47a86a2b53963d9896a7ea437/docs/cs_local_search.ipynb) for searches against an index of size 10k.

My intuition is that this is heading in the right direction but we'll need to experiment with several modifications:

* Experiment with improving the co-training problem - autoencoding instead of error-correcting, code as well as doc strings, code to string and string to code.
* Scope the encoder and decoder variables by modality (separate networks).
* Experiment with a dense layer post-encoder instead of a reduce_mean to yield embedding vector.
* Why is the vocab only ~8k in size? Seems like it should be larger. Confirm this is not an indication of an issue with the dataset.
* Try a larger model or different model hparams: Currently these were trained with more or less the same settings as transformer_tiny.
* Perform a much longer training run as well as eval on validation set to check for overfitting.
* Explore other variants of transpose similarity losses.

And a more involved idea:
* Consider retaining the \n characters in BQ preprocessing step to enable richer embeddings and building AST's; reformulate Problem to produce ASTs from code.

Related to https://github.com/kubeflow/examples/issues/268

cc @activatedgeek @hamelsmu @jlewi

**Comment by jlewi** [comment]
If someone wants to write up a short (~ 1 page doc) summarizing the problem and listing various ideas and things we've tried; I'd be happy to circulate it among the folks here @ Google.

/cc @hamelsmu @cwbeitel @activatedgeek

**Comment by cwbeitel** [comment]
I would be happy to do that or contribute to that.

**Comment by cwbeitel** [comment]
Here's a [doc](https://docs.google.com/document/d/1NmZcZcu9C59ra8j9HLP01wqo9kh9bD9EJdEDRYBzbKE/edit?usp=sharing), let me know if I missed anything @hamelsmu, @activatedgeek, or @jlewi.

**Comment by hamelsmu** [comment]
cc: @rand

**Comment by cwbeitel** [comment]
That sounds cool! Nice work 👍

Can you clarify what vector you're referring to? Is this an embedding
vector or an integer encoded name string? Also can you also clarify what
you mean by a reference vector? Perhaps you can you clarify more what the
precise steps are you're describing here:

"To fix this, we ran the networks on a select batch before feeding it to
the twin/triplet networks, adn then sorting the batches so that the samples
would always be comprised of..."

Nice to hear a fresh idea!

**Comment by cwbeitel** [comment]
@AlexanderCHarrington just @mentioning you in case you're like me and only
see these if you're mentioned!

**Comment by AlexanderCHarrington** [comment]
Thank you!

I am sorry for being so brief and leaving so many things unexplained.

**For the vectors**, I am referring to the encoded strings. I haven't looked at all the intricacies of your model and the surrounding code, _but from what I gathered_:

- **You're comparing two batches of strings with eachother**, effectively in a twin NN architecture (I've used up to 3 networks, i.e. triplets), and calculate an all-to-all distance between the two batches. Each batch represents a document of code, and you sample them in  a dissimilar way. For nomenclature, we say that a **"reference sample"** document holds code for project A, and that the **"negative sample"** document holds code for project B.
Because they're not the same, their average similarity (if we use a l2-normalized dot-product) should be --> -1, and their average distance (l2-normed dot/cosine, MAE, MSE, etc.) should be --> 0 across the two batches.

- **You're experimenting with / suggesting internal distances as well**, since you know that the strings within document A should all output similar vectors since they're from the same underlying codebase. This means that you can express a combined loss as something like this:
E = -sim(A,A), +sim(A,B) _(and possibly -sim(B,B)_

**My experiments**: I've tried to do the exact same thing for both art-recommendations, where the reference sample, A, is an aggregate of a users interactions and a negative sample, B, is an aggregate of randomly sampled actions that the user has not had. I've also tried to do it for entity recognition; A is a reference set of aliases for the same underlying entity, while B is a negative set of aliases for an entity that is not equal to that of A. Even with different loss functions, I have had the same symptoms.

**The symptons**: Loss shoots to very small values, yet when you run predictions, there is no real value in the encoded vectors; the internal similarity matrix for all-to-all vectors across every single piece of string - similar or dissimilar - averages to 1. In effect, the network has bundled everything together in one big ball of yarn, even though the loss says otherwise.

**Some notes for your project**: I saw a brief discussion of output activation and more. I'd personally just use a linear output activation and constrain it with l2 normalization. When I finally get twin/triplet architectures to work, this output is by far the best and easiest to optimize.

_**To be continued**_

**Comment by cwbeitel** [comment]
@AlexanderCHarrington Oh no worries thanks for the extensive follow-up. That's interesting you mention all examples in a batch originating from the same document. That's not what we're doing now - functions and their doc-strings stay paired but functions across the whole data set get shuffled.

**Comment by AlexanderCHarrington** [comment]
@cwbeitel, alright.

1: So you are trying to pair a doc-string with a positive function example and a doc-string with a randomly shuffled negative sample?

So
sim(a_docstring, a_funcstring) --> 1
&
sim(a_docstring, rand_funcstring) --> -1
?

2: I see that you generally discuss the weird loss history, but are there similarities with my symptoms? More specifically, when you measure similarity scores out of sample, performance is incredibly poor, or measured distances don't align with what your loss would have you expect?

**_Note**:_
